src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            /// <summary>
+            /// Starts the wait thread.
+            /// </summary>
+            public void Start()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                waitThread.IsBackground = true;
+                waitThread.Start();
+            }
+
+            /// <summary>
+            /// Queues on the thread pool or executes directly a call to <see cref="UnregisterWait(object)"/>.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in an invalid <see cref="WaitHandle"/>
+            /// into <see cref="RegisteredWait.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the thread pool.
+            /// </remarks>
+            public void QueueOrExecuteUnregisterWait(RegisteredWait handle)
+            {
+                if (handle.IsBlocking)
+                {
+                    UnregisterWait(handle);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(UnregisterWait, handle);
+                }
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="state">The wait handle to unregister.</param>
+            private void UnregisterWait(object state)
+            {
+                RegisteredWait handle = (RegisteredWait)state;
+
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    Debug.Assert(Array.IndexOf(_registeredWaits, handle) != -1 || handle.IsUnregistered);
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_pendingRemoves, handle) == -1 && !handle.IsUnregistered)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                handle.SignalUserWaitHandle();
```

Comment:
There would still be a gap between WaitAny completing and queuing during which an unregister could successfully happen. I think a decent way to do this is to have the wait thread / callback thread signal the event provided to Unregister, since they know precisely when to signal. I'm thinking something along the lines of this:

```c#
        // WaitThreadStart, before any QueueWaitCompletion
        RegisteredWait rw;
        rw.Lock.Acquire();
        try
        {
            ++rw.CallbackCount;
        }
        finally
        {
            rw.Release();
        }

        // ProcessRemovals, before removing a wait
        RegisteredWait rw = _registeredWaits[j];
        rw.Lock.Acquire();
        try
        {
            if (rw.UserUnregisterWaitHandle != null && !rw.UserUnregisterWaitHandleValue != -1)
            {
                if (rw.CallbackCount == 0)
                {
                    rw.SignalUserWaitHandle();
                    // Above is equivalent to:
                    //   if (Interlocked.Exchange(ref _unregisterSignaled, 1) == 0)
                    //   {
                    //       WaitHandle.Set(UserUnregisterWaitHandle);
                    //   }
                }
                else
                {
                    rw.NeedToSignalUserWaitHandleAfterCallbacks = true; // better name?
                }
            }
        }
        finally
        {
            rw.Release();
        }

        // CompleteWait, after the callback returns
        RegisteredWait rw = handle.CompletedHandle;
        rw.Lock.Acquire();
        try
        {
            --rw.CallbackCount;
            if (rw.CallbackCount == 0 && rw.NeedToSignalUserWaitHandleAfterCallbacks)
            {
                rw.SignalUserWaitHandle();
                // Above is equivalent to:
                //   if (Interlocked.Exchange(ref _unregisterSignaled, 1) == 0)
                //   {
                //       WaitHandle.Set(UserUnregisterWaitHandle);
                //   }
            }
        }
        finally
        {
            rw.Release();
        }
```

All the while, Unregister does not have to do any waiting, so it would not have to queue a background work item, and it can return immediately after recording the user event and registering the removal. Does this work?